<b>Привет, Имран!</b>

Меня зовут Екатерина, и я буду проверять твой проект. Давай общаться на «ты», но если это неудобно, пожалуйста, сообщи. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе совершенствоваться как профессионалу.

Ты подготовил замечательный проект, но есть аспекты, которые еще можно улучшить. Я буду использовать комментарии с цветовой разметкой. Пожалуйста, **не перемещай, не изменяй и не удаляй их**. Это поможет выполнить повторную проверку твоего проекта оперативнее.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Таким комментарием выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> ⚠️
    
Таким комментарием выделены небольшие замечания и рекомендации, которые будет хорошо учесть.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Таким комментарием выделены удачные решения🙂
</div>

Свои комментарии ты можешь помечать синим цветом.

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Так ты можешь оформить свой комментарий
</div>

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. После того, как ты все доработаешь, я смогу принять твой проект!) 

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Екатерина, привет! Конечно, можем на "ты"

Удалил
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Давай посмотрим, что получилось)
</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span><ul class="toc-item"><li><span><a href="#Описываем-алгоритм" data-toc-modified-id="Описываем-алгоритм-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Описываем алгоритм</a></span></li><li><span><a href="#Создаем-класс-для-алгоритма" data-toc-modified-id="Создаем-класс-для-алгоритма-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Создаем класс для алгоритма</a></span></li><li><span><a href="#Итоги-по-алгоритму-преобразования" data-toc-modified-id="Итоги-по-алгоритму-преобразования-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Итоги по алгоритму преобразования</a></span></li></ul></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Подготовим-данные-для-модели" data-toc-modified-id="Подготовим-данные-для-модели-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Подготовим данные для модели</a></span></li><li><span><a href="#Проверяем-алгоритм" data-toc-modified-id="Проверяем-алгоритм-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Проверяем алгоритм</a></span></li><li><span><a href="#Рассчитываем-R2-на-исходных-данных" data-toc-modified-id="Рассчитываем-R2-на-исходных-данных-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Рассчитываем R2 на исходных данных</a></span></li><li><span><a href="#Рассчитываем-R2-на-преобразованных-данных" data-toc-modified-id="Рассчитываем-R2-на-преобразованных-данных-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Рассчитываем R2 на преобразованных данных</a></span></li><li><span><a href="#Подводим-итог-проверки" data-toc-modified-id="Подводим-итог-проверки-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Подводим итог проверки</a></span></li></ul></li><li><span><a href="#Итоги-по-проекту" data-toc-modified-id="Итоги-по-проекту-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Итоги по проекту</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Хорошее начало! С содержанием ориентироваться в проекте намного проще👍 Можно добавить описание данных
</div>

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> ⚠️
    
Если используешь собственный класс линейной регрессии, то импортировать из sklearn его не нужно
</div>

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Согласен, но на момент импорта я еще не определился с методом решения задачи)

Удалил
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
👍
</div>

In [2]:
data = pd.read_csv('/datasets/insurance.csv')

In [3]:
display(data.info())
display(data.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


None

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

Данные загружены, что можно сказать:
1. Датасет состоит из 5 столбцов и 5000 строк. Пропуски в данных отсутствуют.
2. К формату столбцов критических замечаний нет.

Далее проведем проверку на дубликаты, в данном случае можно обойтись общей проверкой:

In [4]:
data.duplicated().sum()

153

По итогам проверки выявлены 153 дубликата.
Для принятия решения по их удалению данных недостаточно - в датасете нет ни одного идентификационного поля, по которому можно было бы опрееделить клиента, а среди представленных параметров вполне могут быть дубликаты. Поэтому оставляем как есть.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Отлично, данные загружены и изучены! Можно еще посмотреть на статистики по столбцам `describe()`
</div>

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Так точно, как правило использую связку describe + boxplot, но в конкретно этой задаче я опустил многие проверки данных, чтобы не перегружать проект (при условии, что это все не влияет на его результат)

</div>

## Умножение матриц

Здесь нам необходимо ответить на вопрос: Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново.)

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

**Ответ:** Нет

**Обоснование:** 

1) Для начала приведем наши формулы в рамках задачи:


Предсказания:

$$
a = Xw
$$


Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$


В нашем случае, при умножении на обратимую матрицу формулы примут вид (добавление ..._c означает операции с перемноженной матрицей):

$$
a_c = (XP)w_c
$$


$$
w_c = (X_c^T X_c)^{-1} X_c^T y
$$

Преобразование №1:
$$
w_c = ((XP)^T (XP))^{-1} (XP)^T y
$$

Преобразование №2:
$$
w_c = (P^TX^TXP)^{-1} P^TX^T y
$$

Преобразование №3:
$$
w_c = P^{-1}(X^TX)^{-1} (P^T)^{-1}P^TX^Ty
$$

Преобразование №4:
$$
w_c = P^{-1}(X^TX)^{-1} EX^Ty
$$

Преобразование №5:
$$
w_c = P^{-1}(X^TX)^{-1} X^Ty
$$

Преобразование №6:
$$
w_c = P^{-1}w
$$

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Преобразование №2 нужно проверить. Обрати внимание на порядок множителей. Это важно, поскольку операция умножения двух матриц выполнима только в том случае, если число столбцов в первом сомножителе равно числу строк во втором. Нужно использовать свойство матриц $(AB)^T=B^T A^T$

</div>

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Поправил преобразование №2.

Кстати, если есть какие-нибудь рекомендации доп. материалов по теме - буду очень рад. Я проходил через эту тему с некоторым скрипом, хотелось бы закрепить.
    
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Оставлю полезную ссылку на теорию по обратным матрицам: https://www.berdov.com/works/matrix/obratnaya-matrica/.
</div>

Итого, $$w_c$$ выражается через $$P, w$$ следовательно, $$a_c$$ можно выразить через $$X, P, w$$

Получаем:
$$
a_c = X_cw_c
$$

$$
a_c = (XP)P^{-1}w
$$

$$
a_c = XPP^{-1}w = XEw = Xw = a
$$

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> ⚠️
    
Здесь можно дописать доказательство равенства предсказаний
    
$a_c = XPP^{-1}w = XEw = Xw = a$
</div>

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Да, получилось обрывочно, дополнил последнюю формулу.
    
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Так как при расчете метрики r2 используются истинный таргет и предсказания моделей, а ни то, ни другое не меняется, то и качество не изменится
</div>

Отсюда исходит предположение, что влияния на качество линейной регрессии быть не должно.

## Алгоритм преобразования

### Описываем алгоритм

**Алгоритм**
В целом, здесь можно пойти двумя путями: написать class для регресии, или же воспользоваться базовой моделью linear regressor.
Попробуем написать class.


**Обучение**
1) Создать переменную self.P со случайной обратимой матрицей, проверить на обратимость.

2) Создать переменную X с произведением обучающей выборки на P.

3) Добавить единичный столбец в X.

4) Создать переменную y, содержащую целевой признак.

5) Создать переменную w, рассчитывающуюся по формуле из прошлого пункта.

6) Создать переменные self.w и self.w0, содержащие значения из переменной w.

7) Реализовать механизм включения защиты данных в виде параметра.

**Предсказание**
1) Создать переменную X с произведением тестовой выборки на self.P.

2) Реализовать механизм определения данных - с защитой или без.

3) Вернуть результаты предсказания = X * self.w + self.w0.

**Обоснование**

Для выполнения задачи, нам необходимо рассчитать метрику r2 на исходных данных и защищенных данных. Вышеописанный алгоритм позволяет выполнить работу модели в обоих случаях - благодаря дополнительному параметру включения защиты данных, а также соответствующих доработок в функциях обучения и предсказаний.

### Создаем класс для алгоритма

In [5]:
class LinearRegression:
    
    def protect(self, protection):
        if protection == True:
            self.protect = True
        else:
            self.protect = False
    
    def fit(self, train_features, train_target):
        d = 0
        while d == 0:
            self.P = np.random.normal(size=(len(train_features.columns), len(train_features.columns)))
            d = np.linalg.det(self.P) 
        if self.protect == True:
            X = train_features.dot(self.P)
            X = np.concatenate((np.ones((X.shape[0], 1)), X), axis=1)
        else:
            X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        w = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)
        self.w = w[1:]
        self.w0 = w[0]
    
    def predict(self, test_features):
        if self.protect == True:
            X = test_features.dot(self.P)
        else:
            X = test_features
        return X.dot(self.w) + self.w0

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Хорошо, что проверяешь матрицу $P$ на обратимость
</div>

### Итоги по алгоритму преобразования

Описан и реализован алгоритм преобразования путем создания отдельного класса регрессии, с функционалом, позволяющим работать как исходными данными, так и выполнить защиту данных и произвести расчеты уже на защищенных данных.

## Проверка алгоритма

### Подготовим данные для модели

In [6]:
features = data.drop(['Страховые выплаты'], axis = 1)
target = data['Страховые выплаты']
display(features.shape, target.shape)

(5000, 4)

(5000,)

In [7]:
train_features, test_features, train_target, test_target = train_test_split(
    features, target, test_size = 0.25, random_state = 12345) 
display(train_features.shape, test_features.shape, train_target.shape, test_target.shape)

(3750, 4)

(1250, 4)

(3750,)

(1250,)

Выборки готовы для работы модели.
В данном случае, сознательно не применяю стандартизацию данных, т.к. в контексте задачи она не имеет никакого влияния и смысла.

### Проверяем алгоритм

### Рассчитываем R2 на исходных данных

In [8]:
model = LinearRegression()
model.protect(False)
model.fit(train_features, train_target)
predictions = model.predict(test_features)
r2_not_protect = r2_score(test_target, predictions)
print('Not protect r2_score:', r2_not_protect)

Not protect r2_score: 0.4352275712702667


### Рассчитываем R2 на преобразованных данных

In [9]:
model = LinearRegression()
model.protect(True)
model.fit(train_features, train_target)
predictions = model.predict(test_features)
r2_protect = r2_score(test_target, predictions)
print('Protected r2_score:', r2_protect)

Protected r2_score: 0.43522744229252475


### Подводим итог проверки

Выведем отдельно метрики и разность между ними:

In [10]:
print('R2 на исходных данных:', r2_not_protect)
print('R2 на защищенных данных:', r2_protect)
print('Разница между метриками:', r2_not_protect - r2_protect)

R2 на исходных данных: 0.4352275712702667
R2 на защищенных данных: 0.43522744229252475
Разница между метриками: 1.2897774193110223e-07


Разница между r2 стремится к 0, предполагаю, что здесь можно сделать вывод об отсутствии разницы в работе модели на исходных данных и защищенных данных.

## Итоги по проекту

Итого, выполнено по проекту:
1. Загружены необходимые библиотеки и данные. Данные изучены: на вход получили датасет из 5000 строк и 5 колонок, которые содержат параметры для модели. Пропусков не обнаружено. Дубликаты имеются, но, с учетом отсуствия каких-либо колонок с идентификационными данными, дубликаты решено оставить.

2. Дан ответ с обоснованием на вопрос: "Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново.)". Ответ - не изменится.

3. Описан и реализован алгоритм преобразования путем создания отдельного класса регрессии, с функционалом, позволяющим работать как исходными данными, так и выполнить защиту данных и произвести расчеты уже на защищенных данных.

4. Подведены итоги испытания алгоритма:\
R2 на исходных данных: 0.4352275712702667\
R2 на защищенных данных: 0.4352271295541461\
Разница между метриками: 4.417161205694242e-07\

Разница между r2 стремится к 0, поэтому можно сделать вывод, что алгоритм работает корректно, метрики не имеют значимых отличий.

<div class="alert alert-block alert-success">
    
**Комментарий ревьюера №1** ✅

Верно, качество предсказания действительно не изменяется. Незначительное различие вызвано особенностями преобразования матриц и чисел с плавающей точкой. В случае домножения матрицы признаков на случайную обратимую матрицу можно обеспечить защиту данных. Можно также попробовать декодировать данные обратно.
    
Числовые значения лучше в выводе не выписывать, поскольку при каждом перезапуске они будут отличаться. Либо же можно установить сид рандома
</div>

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
В целом, работа хорошая. Ты молодец! Остались небольшие доработки. Нужно поправить теоретическое доказательство. Доделывай и присылай на повторную проверку.

Жду твоих исправлений 😊
</font></b>
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
    
Отлично! Проект принят! 👏 Поздравляю и желаю удачи!
    
    
Оставлю несколько дополнительных материалов. Обзоры по применению криптографии в машинном обучении: https://arxiv.org/pdf/1902.04109.pdf, https://www.researchgate.net/publication/342746786_Some_Applications_of_Machine_Learning_in_Cryptography
Подробнее про аналогичный алгоритм: https://www.cs.cmu.edu/~rjhall/JOS_revised_May_31a.pdf
</font></b>
</div>